File: src/tenant.py
"""Sutando — Per-Tenant Identity (Python twin of src/tenant.ts).

Mirrors the TypeScript module shipped in PR #748. Both languages implement
the same design — the layout, fallback ordering, and validation rules are
identical, and changes must land in both to stay aligned. See `src/tenant.ts`
for the parallel implementation.

Three-level path model:

    $SUTANDO_PRIVATE_DIR/
      ├── tenant-default/                ← today (single-tenant)
      │   ├── machine-<host>/            ← per-device (scope='device')
      │   └── ...                        ← fleet-shared (scope='shared')
      └── tenant-acme/                   ← future second tenant
          └── ...

Tenant id resolution (high → low):
    1. SUTANDO_TENANT_ID env var.
    2. <workspace>/state/tenant.json   {"id": "...", "name": "..."}
    3. 'default'.

`tenant_path(filename, scope, workspace=None, for_tenant=None)` builds the
three-level path with legacy-pre-migration fallback when the tenant
directory doesn't exist yet.
"""
from __future__ import annotations

import json
import os
import re
import socket
import sys
from pathlib import Path
from typing import Literal, Optional

TENANT_ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9_-]{0,63}$")

_cached_tenant_id: Optional[str] = None


def _workspace_dir() -> Path:
    """Mirrors src/util_paths.py:REPO_DIR + SUTANDO_WORKSPACE env override."""
    ws = os.environ.get("SUTANDO_WORKSPACE")
    if ws:
        return Path(os.path.expanduser(ws))
    return Path(__file__).resolve().parent.parent


def _read_from_state_file(workspace: Path) -> Optional[str]:
    path = workspace / "state" / "tenant.json"
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text())
        tid = data.get("id") if isinstance(data, dict) else None
        if isinstance(tid, str) and TENANT_ID_PATTERN.match(tid):
            return tid
        if tid is not None:
            print(f"[Tenant] state/tenant.json id {tid!r} failed validation; using default", file=sys.stderr)
    except Exception as e:
        print(f"[Tenant] failed to read state/tenant.json: {e}", file=sys.stderr)
    return None


def tenant_id(workspace: Optional[Path] = None) -> str:
    """Resolve the current tenant id. Cached after first call.

    Call `_reset_for_tests()` in tests that switch IDs.
    """
    global _cached_tenant_id
    if _cached_tenant_id is not None:
        return _cached_tenant_id
    ws = workspace if workspace is not None else _workspace_dir()
    env_id = os.environ.get("SUTANDO_TENANT_ID")
    if env_id and TENANT_ID_PATTERN.match(env_id):
        _cached_tenant_id = env_id
        return env_id
    if env_id:
        print(f"[Tenant] SUTANDO_TENANT_ID {env_id!r} failed validation; using default", file=sys.stderr)
    from_state = _read_from_state_file(ws)
    if from_state:
        _cached_tenant_id = from_state
        return from_state
    _cached_tenant_id = "default"
    return "default"


PathScope = Literal["device", "shared"]


def tenant_path(
    filename: str,
    scope: PathScope = "shared",
    workspace: Optional[Path] = None,
    for_tenant: Optional[str] = None,
) -> Path:
    """Per-tenant path resolver. See module docstring for the layout.

    Workspace fallback (mirrors existing personal_path semantics): if
    SUTANDO_PRIVATE_DIR is unset OR the resolved path doesn't exist,
    returns `<workspace>/<file>` for single-tenant installs that haven't
    migrated yet. Callers do `.exists()` checks against the return.

    Returns the FIRST existing path along the priority chain. When nothing
    exists, returns the *preferred* private-dir path so the caller's
    `.exists()` check fails gracefully and a subsequent write lands there.

    `for_tenant`: cross-tenant queries pass it explicitly; default uses the
    current cached tenant id. Keeping both shapes lets a caller read
    another tenant's data without invalidating their own cache.
    """
    ws = workspace if workspace is not None else _workspace_dir()
    tid = for_tenant if for_tenant is not None else tenant_id(ws)
    private_root_env = os.environ.get("SUTANDO_PRIVATE_DIR")
    host = socket.gethostname().split(".")[0]

    if private_root_env:
        root = Path(os.path.expanduser(private_root_env))
        tenant_root = root / f"tenant-{tid}"
        candidate = (
            tenant_root / f"machine-{host}" / filename
            if scope == "device"
            else tenant_root / filename
        )
        if candidate.exists():
            return candidate
        # Single-tenant pre-migration fallback: an existing install may
        # still have files at $SUTANDO_PRIVATE_DIR/machine-<host>/<file>
        # (no tenant-default prefix). Try that path before workspace.
        if tid == "default":
            legacy = (
                root / f"machine-{host}" / filename
                if scope == "device"
                else root / filename
            )
            if legacy.exists():
                return legacy

    ws_path = ws / filename
    if ws_path.exists():
        return ws_path

    if private_root_env:
        root = Path(os.path.expanduser(private_root_env))
        tenant_root = root / f"tenant-{tid}"
        return (
            tenant_root / f"machine-{host}" / filename
            if scope == "device"
            else tenant_root / filename
        )
    return ws_path


def _reset_for_tests() -> None:
    """Test-only: clear the tenant-id cache between tests."""
    global _cached_tenant_id
    _cached_tenant_id = None

File: src/tenant.ts
/**
 * Sutando — Per-Tenant Identity
 *
 * Implements §5 Next: "Per-tenant identity scaffold. Introduce `tenant_id`
 * in path resolution. Single-tenant for now, but the resolver becomes
 * three-level (per-tenant / per-device / fleet-shared). Unblocks any
 * non-owner deployment."
 *
 * The three-level model:
 *
 *   $SUTANDO_PRIVATE_DIR/
 *     ├── tenant-default/                 ← today (single-tenant)
 *     │   ├── machine-<host>/             ← per-device (existing personalPath)
 *     │   │   ├── stand-identity.json
 *     │   │   └── pending-questions.md
 *     │   ├── notes/                      ← fleet-shared (existing sharedPersonalPath)
 *     │   └── build_log.md
 *     └── tenant-acme/                    ← future second tenant
 *         ├── machine-<host>/
 *         └── ...
 *
 * Today's single-tenant install has no `tenant-default/` prefix on disk
 * (the existing `personalPath` writes directly to `$SUTANDO_PRIVATE_DIR/
 * machine-<host>/`). This scaffold introduces the resolver function but
 * does NOT migrate existing call sites — that's a deferred follow-up.
 * `tenantPath()` is the new opt-in entry point; `personalPath` /
 * `sharedPersonalPath` keep working unchanged.
 *
 * Migration plan (separate PR after this lands):
 *   1. Backfill `$SUTANDO_PRIVATE_DIR/tenant-default/` symlinked to the
 *      existing layout (or moved). The existing personalPath() falls back
 *      to the workspace anyway, so a one-time move + symlink would migrate.
 *   2. Switch personalPath/sharedPersonalPath to delegate to tenantPath().
 *   3. Single owner becomes "the default tenant"; the engine is ready for
 *      a second tenant without sprinkling `tenant_id` through 30 callers.
 *
 * Tenant ID resolution (high → low precedence):
 *   1. `SUTANDO_TENANT_ID` env var (deployment knob).
 *   2. `<workspace>/state/tenant.json` { "id": "...", "name": "..." }.
 *   3. Falls back to `"default"` (today's single-tenant reality).
 *
 * NOT in scope here:
 *   - Tenant-scoped access policy / auth — that's the §5 Next/Later
 *     "Per-device access policy" + auth boundaries.
 *   - Tenant directory provisioning — owner creates the directory.
 *   - Cross-tenant audit isolation — `audit.jsonl` per tenant is a
 *     follow-up.
 */

import { existsSync, readFileSync } from 'node:fs';
import { hostname } from 'node:os';
import { join } from 'node:path';

const TENANT_ID_PATTERN = /^[a-z0-9][a-z0-9_-]{0,63}$/;

let cachedTenantId: string | null = null;

function expandHome(p: string): string {
	return p.replace(/^~/, process.env.HOME || '');
}

// Memoize the short hostname — called 3× per tenantPath() invocation,
// and tenantPath() fires on every personalPath/sharedPersonalPath call
// after the #753 migration. Hostname doesn't change at runtime; cache.
const HOSTNAME_SHORT = hostname().split('.')[0];

// Track filenames where we've already warned about both new + legacy
// paths existing simultaneously. Without de-dup, every existsSync
// check during partial migration would log a warning — owner would
// see N×M repeated lines for N readers × M filenames.
const _partialMigrationWarned = new Set<string>();

function workspaceDir(): string {
	return process.env.SUTANDO_WORKSPACE || process.cwd();
}

function readFromStateFile(workspace: string): string | null {
	const path = join(workspace, 'state', 'tenant.json');
	if (!existsSync(path)) return null;
	try {
		const data = JSON.parse(readFileSync(path, 'utf-8')) as { id?: string };
		if (typeof data.id === 'string' && TENANT_ID_PATTERN.test(data.id)) {
			return data.id;
		}
		console.warn(`[Tenant] state/tenant.json id "${data.id}" failed validation; using default`);
	} catch (err) {
		console.warn(`[Tenant] failed to read state/tenant.json: ${(err as Error).message}`);
	}
	return null;
}

/** Resolve the current tenant id. Cached after first call to avoid disk
 *  hits in the hot path. Call `__resetForTests()` in tests that switch IDs. */
export function tenantId(workspace?: string): string {
	if (cachedTenantId !== null) return cachedTenantId;
	const ws = workspace ?? workspaceDir();
	const envId = process.env.SUTANDO_TENANT_ID;
	if (envId && TENANT_ID_PATTERN.test(envId)) {
		cachedTenantId = envId;
		return envId;
	}
	if (envId) {
		console.warn(`[Tenant] SUTANDO_TENANT_ID "${envId}" failed validation; using default`);
	}
	const fromState = readFromStateFile(ws);
	if (fromState) {
		cachedTenantId = fromState;
		return fromState;
	}
	cachedTenantId = 'default';
	return 'default';
}

export type PathScope = 'device' | 'shared';

/** Per-tenant path resolver. Three-level structure:
 *
 *  - `device`: `$SUTANDO_PRIVATE_DIR/tenant-<id>/machine-<host>/<file>`
 *  - `shared`: `$SUTANDO_PRIVATE_DIR/tenant-<id>/<file>`
 *
 *  Workspace fallback (mirrors existing personalPath semantics): if
 *  `SUTANDO_PRIVATE_DIR` is unset OR the resolved path doesn't exist,
 *  returns `<workspace>/<file>` for single-tenant installs that haven't
 *  migrated yet. Callers do `existsSync()` checks against the return.
 *
 *  Returns the FIRST existing path along the priority chain. When nothing
 *  exists, returns the *preferred* private-dir path so an existsSync
 *  check fails gracefully and a subsequent write lands in the right place.
 */
export function tenantPath(filename: string, scope: PathScope = 'shared', workspace?: string, forTenant?: string): string {
	const ws = workspace ?? workspaceDir();
	// Cross-tenant queries pass `forTenant` explicitly; default uses the
	// current cached tenant id. Keeping both shapes lets a caller read
	// another tenant's data without invalidating their own cache.
	const id = forTenant ?? tenantId(ws);
	const privateRoot = process.env.SUTANDO_PRIVATE_DIR;

	if (privateRoot) {
		const root = expandHome(privateRoot);
		const tenantRoot = join(root, `tenant-${id}`);
		const candidate = scope === 'device'
			? join(tenantRoot, `machine-${HOSTNAME_SHORT}`, filename)
			: join(tenantRoot, filename);
		// Single-tenant pre-migration fallback: an existing install may
		// still have files at $SUTANDO_PRIVATE_DIR/machine-<host>/<file>
		// (no tenant-default prefix). Try that path too before workspace.
		const candidateExists = existsSync(candidate);
		if (id === 'default') {
			const legacy = scope === 'device'
				? join(root, `machine-${HOSTNAME_SHORT}`, filename)
				: join(root, filename);
			const legacyExists = existsSync(legacy);
			// Partial-migration warn: both paths exist for the same filename
			// → owner has moved some files but not all. Resolver picks the
			// new tenant path silently; warn-once so owner notices the
			// inconsistency. Cheap (one log per filename per process).
			if (candidateExists && legacyExists && !_partialMigrationWarned.has(filename)) {
				_partialMigrationWarned.add(filename);
				console.warn(
					`[Tenant] partial migration: '${filename}' exists at BOTH ` +
					`'${candidate}' AND legacy '${legacy}'. Resolver picks the tenant path. ` +
					`Move or remove one to silence this warning.`,
				);
			}
			if (candidateExists) return candidate;
			if (legacyExists) return legacy;
		} else {
			if (candidateExists) return candidate;
		}
	}

	const wsPath = join(ws, filename);
	if (existsSync(wsPath)) return wsPath;

	// Nothing exists; return preferred path so the caller's existsSync
	// check fails gracefully and a write lands in the right place.
	if (privateRoot) {
		const root = expandHome(privateRoot);
		const tenantRoot = join(root, `tenant-${id}`);
		return scope === 'device'
			? join(tenantRoot, `machine-${HOSTNAME_SHORT}`, filename)
			: join(tenantRoot, filename);
	}
	return wsPath;
}

/** @internal — test-only reset for the tenant-id cache + partial-migration warn dedup. */
export function __resetForTests(): void {
	cachedTenantId = null;
	_partialMigrationWarned.clear();
}
